src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -1,99 +1,327 @@
 package org.broadinstitute.hellbender.tools.spark.sv;
 
 import com.google.common.annotations.VisibleForTesting;
-import htsjdk.samtools.util.Histogram;
-import htsjdk.samtools.SAMReadGroupRecord;
 import htsjdk.samtools.SamPairUtil;
+import htsjdk.samtools.SAMFileHeader;
 import htsjdk.samtools.metrics.MetricsFile;
-import org.apache.spark.api.java.JavaRDD;
-import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics;
-import org.broadinstitute.hellbender.tools.picard.analysis.InsertSizeMetrics;
+import htsjdk.samtools.util.Histogram;
 import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.metrics.MetricAccumulationLevel;
+import org.broadinstitute.hellbender.tools.picard.analysis.InsertSizeMetrics;
 
+import org.apache.spark.api.java.JavaRDD;
+
+import scala.Tuple2;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.StreamSupport;
+import java.util.stream.Collectors;
 import java.io.Serializable;
-import java.util.List;
-import java.util.Map;
-import java.util.TreeMap;
-import java.util.SortedMap;
 
 /**
- * Worker class to collect insert size metrics.
+ * Worker class to collect insert size metrics, add metrics to file, and provides accessors to stats of groups of different level.
+ * TODO: is it desired that mapping quality is collected as well?
+ * TODO: consider refactoring code for other metric collection usages
  */
-public final class InsertSizeMetricsCollectorSpark implements Serializable{
+public final class InsertSizeMetricsCollectorSpark implements Serializable {
     private static final long serialVersionUID = 1L;
 
-    private final InsertSizeMetrics metrics = new InsertSizeMetrics();
-    private Histogram<Integer> htslibHistogram;
+    // stats are null if not requested to be collected
+    private Map<GroupMetaInfo, Map<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>>> statsOfReadGroups = null;
+    private Map<GroupMetaInfo, Map<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>>> statsOfLibraries  = null;
+    private Map<GroupMetaInfo, Map<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>>> statsOfSamples    = null;
+    private Map<GroupMetaInfo, Map<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>>> statsOfAllReads   = null;
```

Comment:
Why not combine these into a single object by putting all four of these maps as values into a map with a MetricAccumulationLevels as keys? Then you could also just pass that map around as method parameters instead of having class fields -- you'd remove statefullness from your method calls that way, which is a good thing.